File: electron/main/providers/local/watch.ts
/**
 * Layer 2 (local) — the local filesystem watch *mechanism*. Policy-driven: what
 * to watch/exclude and which paths are signals come from the shared watch
 * policy (`deriveWatchSpec`), never from inline literals here. Raw paths are fed
 * into the shared ingest (Layer 3), which normalizes, classifies, and debounces
 * into canonical events. This manager adapts the canonical event back to the
 * current `WatchEvent` contract (rel paths) for the provider handler; T5 carries
 * categories through to the renderer.
 *
 * A live session owns exactly one watch (started on connect, stopped when the
 * session ends), so there is no warm/hot pause state: a watch either runs or is
 * torn down. The previous `setPaused` gating was removed with `suspend()`.
 */
import { watch as fsWatch, type FSWatcher as NodeFSWatcher } from 'node:fs';
import { join } from 'node:path';
import { deriveWatchSpec } from '@shared/watch/policy';
import { createGitignoreFilter } from '../../git/gitignoreFilter';
import { createWatchIngest, type WatchIngest } from '../../watch/ingest';
import { createWorkingTreeWatcher, type WorkingTreeWatcher } from './workingTreeWatcher';
import type { WatchEvent, WatchHandler, WatchSubscription } from '../types';

let counter = 0;

interface ActiveWatch {
  workingTree: WorkingTreeWatcher;
  gitWatchers: NodeFSWatcher[];
  ingest: WatchIngest;
}

export class LocalWatchManager {
  private watches = new Map<string, ActiveWatch>();

  constructor(private readonly rootPath: string) {}

  /**
   * Shared scaffolding for both `subscribe()` (project-root-rooted, plus the
   * dedicated git/beads signal watchers below) and `subscribeWorktree()`
   * (a specific worktree root, working-tree-only — no signal watchers): the
   * ingest wiring, the `excludedSegments` derivation, and the working-tree
   * mechanism itself. Registers the entry in `this.watches` keyed by the
   * returned token so both callers' `unsubscribe` shares one teardown path
   * (`unsubscribeOf`).
   */
  private createBase(
    rootPath: string,
    tokenPrefix: string,
    handler: WatchHandler,
  ): { token: string; active: ActiveWatch; feed: (path: string) => void } {
    const token = `${tokenPrefix}-${++counter}`;
    const spec = deriveWatchSpec();
    const gitignored = createGitignoreFilter(rootPath);

    // Ingest owns debounce + classification. We adapt the canonical event back
    // to the WatchEvent contract (rel path strings) the provider handler expects.
    const ingest = createWatchIngest((canonical) => {
      const event: WatchEvent = {
        token,
        paths: canonical.paths.map((p) => p.rel),
        at: canonical.at,
      };
      handler(event);
    });

    const feed = (path: string): void => {
      ingest.feed([path]);
    };

    // Directory-granularity (`.git`, `.beads`) and never-recurse (`node_modules`)
    // segments are dropped from the working-tree mechanism anywhere they appear.
    // For `.beads` this is load-bearing: it must not pin an open FD on `beads.db`
    // (+ -wal/-shm) — blocking `br doctor --repair`'s exclusive lock and
    // contributing to index corruption (local_repo_explorer-fg5z). For `.git` the
    // ref signals need atomic temp+rename detection. Both are watched at directory
    // granularity via the native fs.watch watchers below instead. (The
    // single-handle native recursive path holds no per-file FD, but we keep this
    // exclusion so behavior is identical to the chokidar/Linux path and the
    // dedicated watchers remain the single source of git-state/beads signals.)
    const excludedSegments = new Set<string>([...spec.directoryGranularity, ...spec.neverRecurse]);
    const isExcludedPath = (p: string): boolean =>
      p
        .replace(/\\/g, '/')
        .split('/')
        .some((seg) => excludedSegments.has(seg));

    // Working-tree mechanism: one native recursive fs.watch on macOS/Windows, a
    // chokidar instance on Linux (see workingTreeWatcher.ts). globs is no longer
    // honored — chokidar v4 dropped glob support and callers only ever pass `['.']`.
    const workingTree = createWorkingTreeWatcher({
      rootPath,
      shouldIgnore: (p: string) => gitignored(p) || isExcludedPath(p),
      onPath: feed,
    });

    const active: ActiveWatch = { workingTree, gitWatchers: [], ingest };
    this.watches.set(token, active);
    return { token, active, feed };
  }

  /** Shared teardown for a token registered via `createBase`, regardless of
   *  whether it also picked up dedicated git/beads signal watchers. */
  private unsubscribeOf(token: string): () => Promise<void> {
    return async () => {
      const w = this.watches.get(token);
      if (!w) return;
      this.watches.delete(token);
      w.ingest.dispose();
      for (const gw of w.gitWatchers) {
        try {
          gw.close();
        } catch {
          /* already closed */
        }
      }
      await w.workingTree.close();
    };
  }

  subscribe(_globs: string[], handler: WatchHandler): WatchSubscription {
    const { token, active, feed } = this.createBase(this.rootPath, 'local-watch', handler);

    // `.git` non-recursive: top-level rewrites (HEAD on branch switch, packed-refs
    // on `git pack-refs`). We forward every filename and let the shared policy
    // keep only the git-state signals (HEAD/packed-refs) and drop the rest
    // (index, FETCH_HEAD, lockfiles), so the "what" stays in one place.
    try {
      const w = fsWatch(join(this.rootPath, '.git'), { persistent: true }, (_event, filename) => {
        if (filename) feed(`.git/${filename}`);
      });
      w.on('error', () => {
        /* directory removed / watcher closed; ignore */
      });
      active.gitWatchers.push(w);
    } catch {
      // Non-git workspace or .git/ does not exist; that's fine.
    }

    // `.git/refs` recursive: every ref change (branch/tag/remote, including nested
    // branches). On macOS libuv backs recursive dir watches with FSEvents (path,
    // not inode), so atomic ref renames are detected reliably. Degrades silently
    // where recursive fs.watch is unsupported.
    try {
      const w = fsWatch(
        join(this.rootPath, '.git', 'refs'),
        { persistent: true, recursive: true },
        (_event, filename) => {
          feed(filename ? `.git/refs/${filename}` : '.git/refs');
        },
      );
      w.on('error', () => {
        /* directory removed / watcher closed; ignore */
      });
      active.gitWatchers.push(w);
    } catch {
      // No .git/refs or platform without recursive support; degrade silently.
    }

    // `.beads` non-recursive directory watch. We deliberately do NOT let chokidar
    // descend into `.beads/` (FD-pin hazard above). A directory-level fs.watch
    // reports changes by path without keeping a read handle on the DB file. We
    // forward every filename and let the policy keep only the committed-write
    // signals (`beads.db`, `issues.jsonl`) and drop `-wal`/`-shm` — the latter is
    // critical: a WAL-mode read bumps their mtime and would feed the workgraph's
    // next read in a self-sustaining loop (local_repo_explorer-fg5z).
    try {
      const w = fsWatch(join(this.rootPath, '.beads'), { persistent: true }, (_event, filename) => {
        if (filename) feed(`.beads/${filename}`);
      });
      w.on('error', () => {
        /* .beads/ removed / watcher closed; ignore */
      });
      active.gitWatchers.push(w);
    } catch {
      // No `.beads/` directory; non-beads workspace. That's fine.
    }

    // `.git/worktrees` non-recursive directory watch: a linked worktree being
    // added or removed (`git worktree add`/`remove`) changes this directory's
    // own listing (a `<name>` entry appearing/disappearing). Deliberately NOT
    // recursive — descending into each `<name>/` subdirectory would re-fire on
    // every commit made INSIDE an already-known worktree (its own HEAD/index/
    // logs churn), which is exactly the noise the shared policy's
    // directory-level classification (`classifyWatchPath`) is built to reject.
    // `.git/worktrees` usually does not exist until the first-ever worktree is
    // added, so this watch commonly fails to attach at subscribe time (caught
    // below) — the top-level `.git` watch above already covers that first
    // creation (it's a new entry in `.git`'s own listing); this dedicated watch
    // takes over for every worktree added/removed afterward within the same
    // live session (local_repo_explorer-rc9n).
    try {
      const w = fsWatch(
        join(this.rootPath, '.git', 'worktrees'),
        { persistent: true },
        (_event, filename) => {
          feed(filename ? `.git/worktrees/${filename}` : '.git/worktrees');
        },
      );
      w.on('error', () => {
        /* .git/worktrees/ removed / watcher closed; ignore */
      });
      active.gitWatchers.push(w);
    } catch {
      // No `.git/worktrees/` yet (repo has never had a linked worktree).
    }

    return { token, unsubscribe: this.unsubscribeOf(token) };
  }

  /**
   * Subscribe to filesystem changes rooted at a SPECIFIC worktree path,
   * independent of `subscribe()`'s project-root-rooted watch — the mechanism
   * behind `WorkspaceProvider.subscribeWorktreeWatch` (see that method's doc
   * comment in `../types.ts` for the full working-tree-only contract).
   * Emits paths relative to `worktreePath`, NOT `this.rootPath`. Deliberately
   * does NOT add the dedicated `.git`/`.git/refs`/`.beads`/`.git/worktrees`
   * signal watchers `subscribe()` adds above: a linked worktree's own `.git`
   * is a plain FILE (a `gitdir:` pointer into the primary worktree's
   * `.git/worktrees/<name>`, not a directory), so those git-state/beads
   * signal watches are meaningless here — the project's PRIMARY `subscribe()`
   * watch (shared repo state) already covers them. Still excludes
   * `.git`/`.beads`/`node_modules` SEGMENTS from the working-tree mechanism
   * itself (the same `excludedSegments` derivation `createBase` always
   * applies), so an edit under the worktree's own `.git`-pointer-adjacent
   * paths or a `node_modules` tree is not walked.
   */
  subscribeWorktree(worktreePath: string, handler: WatchHandler): WatchSubscription {
    const { token } = this.createBase(worktreePath, 'local-watch-wt', handler);
    return { token, unsubscribe: this.unsubscribeOf(token) };
  }

  async closeAll(): Promise<void> {
    const all = [...this.watches.values()];
    this.watches.clear();
    for (const w of all) {
      w.ingest.dispose();
      for (const gw of w.gitWatchers) {
        try {
          gw.close();
        } catch {
          /* already closed */
        }
      }
      await w.workingTree.close();
    }
  }
}

File: remote-helper/watch.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"
	gitignore "github.com/sabhiram/go-gitignore"
)

// fallbackDebounce coalesces rapid filesystem events when the spec carries no
// debounce window. The authoritative debounce lives in the TS ingest layer; the
// helper coalesces only to limit RPC chatter over the wire.
const fallbackDebounce = 150 * time.Millisecond

// watchSpec is the serializable "what to watch" policy pushed by the host over
// the watch.subscribe RPC. It is DERIVED from the single TS policy
// (src/shared/watch/policy.ts via deriveWatchSpec); the helper holds no policy
// of its own. defaultWatchSpec() is a fallback used only when a client omits the
// spec (older client / tests); production always sends the derived spec.
type watchSpec struct {
	NeverRecurse         []string `json:"neverRecurse"`
	DirectoryGranularity []string `json:"directoryGranularity"`
	GitStateSignals      []string `json:"gitStateSignals"`
	BeadsSignals         []string `json:"beadsSignals"`
	DebounceMs           int      `json:"debounceMs"`
}

// defaultWatchSpec mirrors src/shared/watch/policy.ts. Fallback only — keep in
// sync with the TS policy, which is canonical. The real client always pushes the
// derived spec, so production behavior is sourced from TS.
func defaultWatchSpec() watchSpec {
	return watchSpec{
		NeverRecurse:         []string{"node_modules"},
		DirectoryGranularity: []string{".git", ".beads"},
		GitStateSignals:      []string{".git/HEAD", ".git/packed-refs", ".git/refs", ".git/worktrees"},
		BeadsSignals:         []string{".beads/beads.db", ".beads/issues.jsonl"},
		DebounceMs:           200,
	}
}

func (s watchSpec) debounce() time.Duration {
	if s.DebounceMs > 0 {
		return time.Duration(s.DebounceMs) * time.Millisecond
	}
	return fallbackDebounce
}

func toSet(values []string) map[string]struct{} {
	set := make(map[string]struct{}, len(values))
	for _, v := range values {
		set[v] = struct{}{}
	}
	return set
}

// relPosix returns the root-relative forward-slash path for abs, plus ok=false
// when abs is outside root.
func relPosix(root, abs string) (string, bool) {
	rel, err := filepath.Rel(root, abs)
	if err != nil || rel == "." || strings.HasPrefix(rel, "..") {
		return "", false
	}
	return filepath.ToSlash(rel), true
}

// gitWorktreesSignal is GIT_STATE_SIGNALS' ".git/worktrees" entry
// (src/shared/watch/policy.ts). Unlike ".git/refs" -- an intentionally
// UNBOUNDED-depth signal, since every nested ref path is real signal -- a
// worktree being added or removed only ever changes ".git/worktrees" itself
// or its immediate "<name>" child. Anything nested deeper
// (".git/worktrees/<name>/HEAD", "/index", "/logs/HEAD", ...) is per-commit
// churn made INSIDE that worktree's own metadata dir, not a worktree-set
// change, and must stay noise -- see GIT_STATE_SIGNALS's doc comment in
// policy.ts for the full rationale.
const gitWorktreesSignal = ".git/worktrees"

// gitWorktreesMaxSegments bounds how many "/"-separated segments a path may
// have and still count as the gitWorktreesSignal: the signal itself
// (".git/worktrees", 2 segments) or exactly one child ("<name>", 3
// segments) -- mirrors policy.ts's classifyWatchPath segment-count gate
// (`segments[1] === 'worktrees' && segments.length <= 3`).
const gitWorktreesMaxSegments = 3

// matchesSignal reports whether rel is one of the signal paths, treating each
// signal as an exact path OR a directory prefix (so ".git/refs" matches
// ".git/refs/heads/main"). gitWorktreesSignal is the one exception: it is
// depth-BOUNDED (see gitWorktreesMaxSegments) rather than an unbounded
// prefix, matching policy.ts's classifyWatchPath segment-count gate.
//
// This bound is Go-side defense-in-depth, independent of
// addWatchesWithSpec's directory-walk `filepath.SkipDir` at ".git/worktrees"
// (which is what actually prevents fsnotify from ever watching deep enough
// to produce these paths in production today). Without this guard, this
// function alone would incorrectly treat a deep per-worktree-metadata path
// as a git-state signal -- currently unreachable dead-code-in-practice, but
// a future unrelated change to the directory walk (e.g. removing that
// SkipDir to watch worktree subdirs for some other feature) could silently
// reintroduce the per-commit noise bug remote-only, with nothing to catch it
// (local_repo_explorer-wkxb).
func matchesSignal(rel string, signals []string) bool {
	for _, sig := range signals {
		if rel == sig {
			return true
		}
		if !strings.HasPrefix(rel, sig+"/") {
			continue
		}
		if sig == gitWorktreesSignal && len(strings.Split(rel, "/")) > gitWorktreesMaxSegments {
			continue // per-worktree metadata churn -- noise, not a signal
		}
		return true
	}
	return false
}

// loadGitignore attempts to compile the root-level .gitignore for the project.
// On any error (missing file, parse error) it returns nil so the watcher
// degrades gracefully.
func loadGitignore(root string) *gitignore.GitIgnore {
	gi, err := gitignore.CompileIgnoreFile(filepath.Join(root, ".gitignore"))
	if err != nil {
		return nil
	}
	return gi
}

// isGitignored reports whether absPath is ignored by gi. Returns false when gi
// is nil (no .gitignore loaded).
func isGitignored(gi *gitignore.GitIgnore, root, absPath string) bool {
	if gi == nil {
		return false
	}
	rel, ok := relPosix(root, absPath)
	if !ok {
		return false
	}
	return gi.MatchesPath(rel)
}

// watcher tracks a single active subscription rooted at a directory.
type watcher struct {
	token  string
	root   string
	gi     *gitignore.GitIgnore // nil when no .gitignore is present
	spec   watchSpec
	fsw    *fsnotify.Watcher
	done   chan struct{}
	closed bool
}

// shouldEmit decides whether a raw filesystem event path is forwarded to the
// host. Emission is driven entirely by the spec, mirroring the TS policy's
// classifyWatchPath:
//   - never-recurse segments (node_modules) are dropped;
//   - paths under a directory-granularity dir (.git/.beads) are emitted only
//     when they match a git-state or beads signal (so .git/index,
//     .beads/beads.db-wal, etc. are dropped — wire churn + self-feed avoidance);
//   - all other paths are emitted unless gitignored.
func (w *watcher) shouldEmit(abs string) bool {
	rel, ok := relPosix(w.root, abs)
	if !ok {
		return false
	}
	segments := strings.Split(rel, "/")
	neverRecurse := toSet(w.spec.NeverRecurse)
	for _, seg := range segments {
		if _, skip := neverRecurse[seg]; skip {
			return false
		}
	}
	dg := toSet(w.spec.DirectoryGranularity)
	if _, isDG := dg[segments[0]]; isDG {
		return matchesSignal(rel, w.spec.GitStateSignals) || matchesSignal(rel, w.spec.BeadsSignals)
	}
	return !isGitignored(w.gi, w.root, abs)
}

// watchManager owns all active watchers and serializes event emission.
type watchManager struct {
	mu        sync.Mutex
	watchers  map[string]*watcher
	emitEvent func(Event)
}

func newWatchManager(emit func(Event)) *watchManager {
	return &watchManager{
		watchers:  make(map[string]*watcher),
		emitEvent: emit,
	}
}

type watchSubscribeParams struct {
	Cwd   string     `json:"cwd"`
	Token string     `json:"token"`
	Spec  *watchSpec `json:"spec,omitempty"`
}

type watchUnsubscribeParams struct {
	Token string `json:"token"`
}

// subscribe begins a watch on cwd, keyed by token. A token already in use is
// replaced. Loads the root .gitignore and applies the pushed watch spec (or the
// default when none is sent).
func (m *watchManager) subscribe(raw json.RawMessage) (interface{}, error) {
	var p watchSubscribeParams
	if err := json.Unmarshal(raw, &p); err != nil {
		return nil, fmt.Errorf("watch.subscribe: decode params: %w", err)
	}
	if p.Cwd == "" {
		return nil, fmt.Errorf("watch.subscribe: cwd must not be empty")
	}
	if p.Token == "" {
		return nil, fmt.Errorf("watch.subscribe: token must not be empty")
	}
	spec := defaultWatchSpec()
	if p.Spec != nil {
		spec = *p.Spec
	}

	m.mu.Lock()
	if existing, ok := m.watchers[p.Token]; ok {
		existing.stop()
		delete(m.watchers, p.Token)
	}
	m.mu.Unlock()

	gi := loadGitignore(p.Cwd)

	fsw, err := fsnotify.NewWatcher()
	if err != nil {
		return nil, fmt.Errorf("watch.subscribe: create watcher: %w", err)
	}

	if err := addWatchesWithSpec(fsw, p.Cwd, p.Cwd, spec, gi); err != nil {
		fsw.Close()
		return nil, fmt.Errorf("watch.subscribe: add %q: %w", p.Cwd, err)
	}

	w := &watcher{
		token: p.Token,
		root:  p.Cwd,
		gi:    gi,
		spec:  spec,
		fsw:   fsw,
		done:  make(chan struct{}),
	}

	m.mu.Lock()
	m.watchers[p.Token] = w
	m.mu.Unlock()

	go m.run(w)

	return map[string]any{"token": p.Token}, nil
}

// unsubscribe stops and removes the watcher for token. Unknown tokens are a
// no-op success.
func (m *watchManager) unsubscribe(raw json.RawMessage) (interface{}, error) {
	var p watchUnsubscribeParams
	if err := json.Unmarshal(raw, &p); err != nil {
		return nil, fmt.Errorf("watch.unsubscribe: decode params: %w", err)
	}
	if p.Token == "" {
		return nil, fmt.Errorf("watch.unsubscribe: token must not be empty")
	}

	m.mu.Lock()
	w, ok := m.watchers[p.Token]
	if ok {
		delete(m.watchers, p.Token)
	}
	m.mu.Unlock()

	if ok {
		w.stop()
	}
	return map[string]any{"token": p.Token}, nil
}

// closeAll stops every active watcher; used on shutdown.
func (m *watchManager) closeAll() {
	m.mu.Lock()
	all := make([]*watcher, 0, len(m.watchers))
	for _, w := range m.watchers {
		all = append(all, w)
	}
	m.watchers = make(map[string]*watcher)
	m.mu.Unlock()
	for _, w := range all {
		w.stop()
	}
}

// run consumes raw fsnotify events for a watcher, debounces them, and emits
// coalesced "watch" events. Errors become "watchError" events.
func (m *watchManager) run(w *watcher) {
	pending := make(map[string]struct{})
	var timer *time.Timer
	var timerC <-chan time.Time
	debounce := w.spec.debounce()

	flush := func() {
		if len(pending) == 0 {
			return
		}
		paths := make([]string, 0, len(pending))
		for p := range pending {
			paths = append(paths, p)
		}
		pending = make(map[string]struct{})
		m.emitEvent(Event{
			Event: "watch",
			Data:  map[string]any{"token": w.token, "paths": paths},
		})
	}

	for {
		select {
		case <-w.done:
			return
		case ev, ok := <-w.fsw.Events:
			if !ok {
				return
			}
			if !w.shouldEmit(ev.Name) {
				// Keep the recursive watch current even for non-emitted dirs (e.g.
				// a new .git/refs/<branch> directory) so its children are caught.
				if ev.Op&fsnotify.Create != 0 {
					if info, statErr := os.Stat(ev.Name); statErr == nil && info.IsDir() {
						_ = addWatchesWithSpec(w.fsw, ev.Name, w.root, w.spec, w.gi)
					}
				}
				continue
			}
			// Emit repo-relative POSIX paths for parity with the local mechanism,
			// so the shared ingest/policy classify both transports identically.
			rel, ok := relPosix(w.root, ev.Name)
			if !ok {
				rel = filepath.ToSlash(ev.Name)
			}
			pending[rel] = struct{}{}
			if ev.Op&fsnotify.Create != 0 {
				if info, statErr := os.Stat(ev.Name); statErr == nil && info.IsDir() {
					_ = addWatchesWithSpec(w.fsw, ev.Name, w.root, w.spec, w.gi)
				}
			}
			if timer == nil {
				timer = time.NewTimer(debounce)
				timerC = timer.C
			} else {
				if !timer.Stop() {
					select {
					case <-timer.C:
					default:
					}
				}
				timer.Reset(debounce)
			}
		case <-timerC:
			flush()
			timer = nil
			timerC = nil
		case err, ok := <-w.fsw.Errors:
			if !ok {
				return
			}
			m.emitEvent(Event{
				Event: "watchError",
				Data:  map[string]any{"token": w.token, "error": err.Error()},
			})
		}
	}
}

// stop closes the watcher and signals its run loop to exit. Safe to call once.
func (w *watcher) stop() {
	if w.closed {
		return
	}
	w.closed = true
	close(w.done)
	_ = w.fsw.Close()
}

// addWatchesWithSpec registers root and the non-excluded subdirectories with the
// watcher, driven by the spec. Directory-granularity dirs (.git/.beads) are
// watched at the directory level (so signal files like .git/HEAD and
// .beads/issues.jsonl are seen) without descending into their heavy/churny
// subtrees — except .git/refs, which is watched recursively for branch/tag
// changes, and .git/worktrees, which is watched at ITS OWN directory level
// only (a linked worktree being added/removed changes that directory's own
// listing) without descending into any individual worktree's metadata dir —
// otherwise a routine commit made inside an already-known worktree (which
// rewrites its own .git/worktrees/<name>/HEAD, /index, /logs/HEAD, …) would
// spam a watch event on every commit, not just on add/remove
// (local_repo_explorer-rc9n). node_modules and gitignored subtrees are pruned
// (EMFILE avoidance). projectRoot is the repository root used for
// gitignore-relative paths.
func addWatchesWithSpec(fsw *fsnotify.Watcher, root, projectRoot string, spec watchSpec, gi *gitignore.GitIgnore) error {
	neverRecurse := toSet(spec.NeverRecurse)
	dg := toSet(spec.DirectoryGranularity)
	return filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil // tolerate transient per-entry errors
		}
		if !d.IsDir() {
			return nil
		}
		if _, skip := neverRecurse[d.Name()]; skip {
			return filepath.SkipDir
		}
		if isGitignored(gi, projectRoot, path) {
			return filepath.SkipDir
		}
		rel, ok := relPosix(projectRoot, path)
		if !ok {
			// path == projectRoot itself: watch it and descend.
			if addErr := fsw.Add(path); addErr != nil {
				return fmt.Errorf("add watch %q: %w", path, addErr)
			}
			return nil
		}
		top := strings.Split(rel, "/")[0]
		if _, isDG := dg[top]; isDG {
			// Watch the .git dir itself and descend only into refs; watch the
			// .beads dir itself and skip its subtrees (signals are direct children).
			if rel == ".git" {
				_ = fsw.Add(path)
				return nil // descend to reach .git/refs
			}
			if strings.HasPrefix(rel, ".git/") {
				if rel == ".git/refs" || strings.HasPrefix(rel, ".git/refs/") {
					_ = fsw.Add(path)
					return nil
				}
				if rel == ".git/worktrees" {
					_ = fsw.Add(path)
					return filepath.SkipDir // per-worktree metadata (HEAD/index/logs) excluded
				}
				return filepath.SkipDir // objects, logs, hooks, info, …
			}
			if rel == ".beads" {
				_ = fsw.Add(path)
				return filepath.SkipDir // .br_history/.br_recovery churn excluded
			}
			if strings.HasPrefix(rel, ".beads/") {
				return filepath.SkipDir
			}
			// Any other directory-granularity dir: watch shallow.
			_ = fsw.Add(path)
			return filepath.SkipDir
		}
		if addErr := fsw.Add(path); addErr != nil {
			return fmt.Errorf("add watch %q: %w", path, addErr)
		}
		return nil
	})
}
